core/src/main/scala/tech/beshu/ror/configuration/EsConfigFileLoader.scala

```diff
@@ -17,26 +17,41 @@
 package tech.beshu.ror.configuration
 
 import better.files.File
+import cats.Show
+import cats.data.NonEmptyList
+import cats.implicits._
 import io.circe.Decoder
+import tech.beshu.ror.accesscontrol.factory.JsonConfigStaticVariableResolver
+import tech.beshu.ror.accesscontrol.factory.JsonConfigStaticVariableResolver.ResolvingError
+import tech.beshu.ror.providers.EnvVarsProvider
 import tech.beshu.ror.utils.yaml
 
-trait LoadFromEsConfig {
+final class EsConfigFileLoader[CONFIG: Decoder]()(implicit envVarsProvider: EnvVarsProvider) {
 
-  protected def loadConfigFromFile[CONFIG: Decoder](file: File,
-                                                    configName: String): Either[MalformedSettings, CONFIG] = {
+  def loadConfigFromFile(file: File,
+                         configName: String): Either[MalformedSettings, CONFIG] = {
     file.fileReader { reader =>
       yaml
         .parser
         .parse(reader)
         .left.map(e => MalformedSettings(s"Cannot parse file ${file.pathAsString} content. Cause: ${e.message}"))
         .right
+        .flatMap { json =>
+          JsonConfigStaticVariableResolver.resolve(json)
+            .left.map(e => MalformedSettings(show"""Invalid $configName configuration. $e."""))
+        }
         .flatMap { json =>
           implicitly[Decoder[CONFIG]]
             .decodeJson(json)
-            .left.map(e => MalformedSettings(s"Invalid $configName configuration"))
+            .left.map(_ => MalformedSettings(s"Invalid $configName configuration"))
         }
     }
   }
+
+  private implicit val showResolvingError: Show[NonEmptyList[ResolvingError]] =
+    Show.show { nel =>
+      nel.map(_.msg).mkString_(", ")
```

Comment:
message in ResolvingError is human readable sentence, so concatenating such sentences with `,` could end up with weird thing. Use multiline string with eg. bullet points

Reply:
Do you mean to use `"\n\n* "` instead of `", "`.

Reply:
seems that `\n* ` will be enough

Reply:
I mean `"\n\t* "` :)

Reply:
👍